### PR TITLE
Smartly handle Shift+Enter better... 

### PR DESCRIPTION
**Pull request by balajikris**
...by treating it as Enter where user can unintentionally have
Shift key pressed while hitting Enter.

In cases of pressing shift+enter after typing open parenthesis to open a block, assuming brace completion is on, we produced code like this:

initial:
```C#
try { |}
```
final:
```C#
try { }
|
```
basically, in such a case, we didn't have anything to complete the line (like a semicolon that we need to insert), yet we handled it anyway and inserted a Newline after the block and positioned caret on the next line.

a better output here could be what we produce on enter:

```C#
try
{
    |
}
```

the implementation is simple: detect cases like this and delegate the command to the editor whose default implementation is exactly what we need.

Fixes internal bug 150480. 
Also this: http://stackoverflow.com/questions/32259287/visual-studio-2015-brackets-not-automatically-formatting-when-pressing-shift

### CODE REVIEW

**Comment by balajikris** [comment]
tagging @dotnet/roslyn-ide for feedback.

@DustinCampbell and @Pilchie already looked at 78d7d93  (via b22c515)

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs [hunk 1]
```diff
@@ -41,6 +41,40 @@ protected override void NextAction(IEditorOperations editorOperation, Action nex
...
+            //      To support this, we treat `shift + enter` like `enter` here.
+            var afterOpenBrace = startToken.Kind() == SyntaxKind.OpenBraceToken
+                              && endToken.Kind() == SyntaxKind.CloseBraceToken
+                              && tokenToLeft == startToken
+                              && endToken.Parent.IsKind(SyntaxKind.Block)
+                              && FormattingRangeHelper.AreTwoTokensOnSameLine(startToken, endToken);
+
+            return afterOpenBrace;
+        }
+
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: Do you need to verify that the position is to the left of the endToken? I seem to recall that FindToken would return the token to the left in this case:

```C#
try {  }|
```

Comment (DustinCampbell) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: Hmmm... maybe not. Looks like your tests cover this case.

File: src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs [hunk 2]
```diff
@@ -51,6 +51,11 @@ internal abstract class AbstractAutomaticLineEnderCommandHandler :
         /// </summary>
         protected abstract void FormatAndApply(Document document, int position, CancellationToken cancellationToken);
 
+        /// <summary>
+        /// special cases where we do not want to do line completion but just fall back to line break and formatting.
+        /// </summary>
+        protected abstract bool TreatAsReturn(Document document, int position, CancellationToken cancellationToken);
```
Comment (dpoeschl) [inline src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs]: Super duper nit: ```TreatAsBreakLine``` (it's not guaranteed to be Return, right?). Or, the negation, ```ShouldCompleteLine```?

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs [hunk 3]
```diff
@@ -41,6 +41,40 @@ protected override void NextAction(IEditorOperations editorOperation, Action nex
...
+            var endToken = root.FindToken(position);
+            if (endToken.IsMissing)
+            {
+                return false;
+            }
+
+            var tokenToLeft = root.FindTokenOnLeftOfPosition(position);
+            var startToken = endToken.GetPreviousToken();
+
+            // case 1:
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: Are there other cases?

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs [hunk 4]
```diff
@@ -41,6 +41,40 @@ protected override void NextAction(IEditorOperations editorOperation, Action nex
             editorOperation.InsertNewLine();
         }
 
+        protected override bool TreatAsReturn(Document document, int position, CancellationToken cancellationToken)
+        {
+            var root = document.GetSyntaxRootAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+
+            var endToken = root.FindToken(position);
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: Personally, I found these names confusing. Maybe "possibleOpenBrace" and "possibleCloseBrace", etc. I understand that you mean "start" and "end" in the typical Formatting Span sense, but I was confused in this context. Maybe get other opinions.

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs (hunk 1 above)
Comment (balajikris) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: Yeah it looks to see if tokenToLeft of caret is an open brace.

File: src/EditorFeatures/Test/AutomaticCompletion/AbstractAutomaticLineEnderTests.cs [hunk 5]
```diff
@@ -44,9 +45,21 @@ protected void Test(string expected, string code, bool completionActive = false)
...
 
-                commandHandler.ExecuteCommand(new AutomaticLineEnderCommandArgs(view, buffer), CreateNextHandler(workspace));
+                commandHandler.ExecuteCommand(new AutomaticLineEnderCommandArgs(view, buffer),
+                                                    assertNextHandlerInvoked
+                                                        ? () => { nextHandlerInvoked = true; }
+                                                        : CreateNextHandler(workspace));
 
                 Test(view, buffer, expected);
+
+                if (assertNextHandlerInvoked)
```
Comment (dpoeschl) [inline src/EditorFeatures/Test/AutomaticCompletion/AbstractAutomaticLineEnderTests.cs]: Equivalent: Assert.Equal(assertNextHandlerInvoked, nextHandlerInvoked)

File: src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs [hunk 6]
```diff
@@ -104,8 +106,21 @@ public void ExecuteCommand(AutomaticLineEnderCommandArgs args, Action nextHandle
...
+                        // VS editor's default implementation of SmartBreakLine is simply BreakLine, which inserts
+                        // a new line and positions the caret with smart indent.
+                        nextHandler();
+                        return;
+                    }
+
                     using (var transaction = args.TextView.CreateEditTransaction(EditorFeaturesResources.AutomaticLineEnder, _undoRegistry, _editorOperationsFactoryService))
                     {
+                        // try to move the caret to the end of the line on which the caret is
+                        args.TextView.TryMoveCaretToAndEnsureVisible(subjectLineWhereCaretIsOn.End);
```
Comment (dpoeschl) [inline src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs]: Unrelated change?

File: src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticLineEnderTests.cs [hunk 7]
```diff
@@ -703,6 +702,144 @@ public void NotAfterAutoPropertyDeclaration()
 }");
         }
 
+        [WorkItem(150480, "devdiv.visualstudio.com")]
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticLineEnderTests.cs]: "devdiv.visualstudio.com" doesn't appear anywhere in the codebase today. Is this how we want to mark internal bugs? (/cc: @Pilchie)

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs (hunk 3 above)
Comment (balajikris) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: Not any more that I know of. Just structured the code to be so that eventually it can `return case1 || case 2 .. `

File: src/EditorFeatures/Test/AutomaticCompletion/AbstractAutomaticLineEnderTests.cs (hunk 5 above)
Comment (balajikris) [inline src/EditorFeatures/Test/AutomaticCompletion/AbstractAutomaticLineEnderTests.cs]: Nice. thanks.

**Comment by dpoeschl** [comment]
:+1:

File: src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticLineEnderTests.cs (hunk 7 above)
Comment (balajikris) [inline src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticLineEnderTests.cs]: I thought I've seen it somewhere? Hmm.. Anyway, with the myriad of bug databases to go about, I thought workitem number is useful only if its qualified with its database name :smile:

**Comment by balajikris** [comment]
got an oral sign off from @Pilchie as he already commented on via  b22c515
